diff --git a/src/intrinsics/Promise.mts b/src/intrinsics/Promise.mts
index 56f5c55..4a10681 100644
--- a/src/intrinsics/Promise.mts
+++ b/src/intrinsics/Promise.mts
@@ -127,6 +127,11 @@ interface PromiseAllAccumulatorFunctionObject extends PromiseAllResolveElementFu
   readonly Capability: PromiseCapabilityRecord;
 }
 
+interface PromiseAnyAccumulatorFunctionObject extends PromiseAllRejectElementFunctionObject {
+  readonly Errors: Value[];
+  readonly Capability: PromiseCapabilityRecord;
+}
+
 function CreatePromiseAllResolveElement(index: number, values: Value[], resultCapability: PromiseCapabilityRecord, remainingElementsCount: { Value: number }): FunctionObject {
   const fulfilledSteps = function* fulfilled([value = Value.undefined]: Arguments): ValueEvaluator {
     const F = surroundingAgent.activeFunctionObject as PromiseAllResolveElementFunctionObject;
@@ -491,11 +496,13 @@ function* PerformPromiseAllSettled(iteratorRecord: IteratorRecord, constructor:
       'Values',
       'Capability',
       'RemainingElements',
-    ])) as Mutable<PromiseAllResolveElementFunctionObject>;
+    ])) as Mutable<PromiseAllAccumulatorFunctionObject>;
 
     const alreadyCalled = { Value: false };
     onFulfilled.AlreadyCalled = alreadyCalled;
     onFulfilled.Index = index;
+    onFulfilled.Values = values;
+    onFulfilled.Capability = resultCapability;
 
     // s. Let rejectedSteps be the algorithm steps defined in Promise.allSettled Reject Element Functions.
     const rejectedSteps = function* PromiseAllSettledRejectElementFunctions([error = Value.undefined]: Arguments): ValueEvaluator {
@@ -519,9 +526,11 @@ function* PerformPromiseAllSettled(iteratorRecord: IteratorRecord, constructor:
     };
 
     // u. Let onRejected be ! CreateBuiltinFunction(rejectedSteps, 1, "", « [[AlreadyCalled]], [[Index]] »).
-    const onRejected = X(CreateBuiltinFunction(rejectedSteps, 1, Value(''), ['AlreadyCalled', 'Index'])) as Mutable<PromiseAllResolveElementFunctionObject>;
+    const onRejected = X(CreateBuiltinFunction(rejectedSteps, 1, Value(''), ['AlreadyCalled', 'Index', 'Values', 'Capability'])) as Mutable<PromiseAllAccumulatorFunctionObject>;
     onRejected.AlreadyCalled = alreadyCalled;
     onRejected.Index = index;
+    onRejected.Values = values;
+    onRejected.Capability = resultCapability;
     index += 1;
     remainingElementsCount.Value += 1;
     Q(yield* Invoke(nextPromise, 'then', [onFulfilled, onRejected]));
@@ -617,9 +626,11 @@ function* PerformPromiseAny(iteratorRecord: IteratorRecord, constructor: Functio
       return Value.undefined;
     };
     // l. Let onRejected be ! CreateBuiltinFunction(stepsRejected, lengthRejected, "", « [[AlreadyCalled]], [[Index]], [[Errors]], [[Capability]], [[RemainingElements]] »).
-    const onRejected = X(CreateBuiltinFunction(rejectedSteps, 1, Value(''), ['AlreadyCalled', 'Index'])) as Mutable<PromiseAllRejectElementFunctionObject>;
+    const onRejected = X(CreateBuiltinFunction(rejectedSteps, 1, Value(''), ['AlreadyCalled', 'Index', 'Errors', 'Capability'])) as Mutable<PromiseAnyAccumulatorFunctionObject>;
     onRejected.AlreadyCalled = { Value: false };
     onRejected.Index = index;
+    onRejected.Errors = errors;
+    onRejected.Capability = resultCapability;
     index += 1;
     remainingElementsCount.Value += 1;
     Q(yield* Invoke(nextPromise, 'then', [resultCapability.Resolve, onRejected]));
